# Computer Lab 4: Modeling El Niño-Southern Oscillation
This notebook will contain the entirety of the El Niño-Southern Oscillation computer lab.

## Setup
For this assignment, we will need `matplotlib.pyplot`, `numpy`, and `scipy.integrate`, as per usual. We will also need specific functions from `sympy` again.

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import scipy.integrate as integrate
from sympy import real_roots, evalf
from sympy.abc import Y

## Section 1: Equilibrium Solutions to the ENSO Model
We begin by looking at equilibrium solutions, as detailed in class.

### Part 1.1
Here, we are asked to seek steady-state solutions to the Vallis ENSO model. Recall that this model, in its simplest form, gives us three equations governing the zonal dynamics of the Pacific as

$$ \frac{d}{dt}U = B(T_E-T_W) + C(U_0-U) $$
$$ \frac{d}{dt}T_E = \frac{U}{\Delta x}T_W + A(T_0-T_E) $$
$$ \frac{d}{dt}T_W = \frac{-U}{\Delta x}T_E + A(T_0-T_W) $$

To find our steady-state solution, we set all of these derivatives equal to zero, and solve. Let's begin with the case where $U_0=-0.5\mathrm{ms}^{-1}$ and $B=1.3\times10^{-7}\mathrm{ms}^{-2}\degree\mathrm{C}^{-1}$. We are also given that $\Delta x = 7500\mathrm{km} = 7.5\times10^6\mathrm{m}$, $T_0=-15\degree\mathrm{C}$, $A = (6\;\mathrm{months})^{-1}$, and $C=(2\;\mathrm{months})^{-1}$. Let's quickly convert $A$ and $C$ to seconds by finding the average number of seconds in a month.

In [4]:
# Year = 365.2425 days x 24 hours x 60 minutes x 60 seconds
sec_year = 365.2425 * 24 * 60 * 60
A = 1 / (6 / 12 * sec_year)
C = 1 / (2 / 12 * sec_year)
print(A, C)

6.337747701362287e-08 1.9013243104086862e-07


We find that $A=6.338\times10^{-8}\mathrm{s}^{-1}$ and $C=1.901\times10^{-7}\mathrm{s}^{-1}$. With that, let's set up our first case and solve.

In [6]:
del_x = 7.5 * 10**6
T_0 = -15
U_0 = -0.5
B = 1.3 * 10**(-7)

def system_eq(vars):
    T_e, T_w, U = vars
    eq_1 = B * (T_e - T_w) + C * (U_0 - U)
    eq_2 = U / del_x * T_w + A * (T_0 - T_e)
    eq_3 = -U / del_x * T_e + A * (T_0 - T_w)

    return [eq_1, eq_2, eq_3]


To solve this system, we will need to use a capability of `scipy` that we haven't before: `scipy.optimize`'s `fsolve()`. Using `fsolve()`, let's assume a good starting estimate for $T_W$ and $T_E$ is $T_0$, and for $U$, $U_0$.

In [11]:
from scipy.optimize import fsolve

T_e, T_w, U = fsolve(system_eq, (T_0, T_0, U_0))
print(T_e, T_w, U)

-14.634128200821218 -15.348835253523902 -0.0113305166156651


In this case, we find that $T_E=-14.624\degree\mathrm{C}$, $T_W=-15.349\degree\mathrm{C}$, and $U=-0.011\mathrm{ms}^{-1}$.